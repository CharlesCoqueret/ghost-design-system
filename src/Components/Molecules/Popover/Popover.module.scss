.container {
  width: fit-content;
  padding: 8px;
  display: flex;
  flex-direction: column;
}

.title {
  margin-bottom: 16px;
  display: flex;
  justify-content: center;
}

.buttons {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;

  & > * {
    margin: auto 4px;
  }
}
