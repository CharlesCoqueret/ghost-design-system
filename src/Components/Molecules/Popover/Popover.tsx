import React, { ReactElement, Ref, useEffect, useRef, useState } from 'react';
import { ControlledMenu } from '@szhsin/react-menu';
import classnames from 'classnames';

import { useOnClickOutside } from '../../../hooks';
import { Portal } from '../../Atoms/Portal';
import Button, { IButtonProps } from '../Button/Button';

export interface IPopoverProps {
  buttons: Array<IButtonProps>;
  open?: boolean;
  anchorRef: React.RefObject<HTMLElement>;
  title?: string;
}

const Popover = (props: IPopoverProps): ReactElement => {
  const { buttons, open, anchorRef, title } = props;

  const [isOpen, setIsOpen] = useState<boolean | undefined>(open);
  const [isShaking, setIsShaking] = useState(false);
  const skipOpen = useRef(false);
  const menuRef = useRef(null);

  const pnClickOutside = () => {
    setIsShaking(true);
    setTimeout(() => {
      setIsShaking(false);
    }, 500);
  };

  useOnClickOutside(menuRef, () => {
    if (open) {
      pnClickOutside();
    }
  });

  useEffect(() => {
    if (!skipOpen.current) {
      setIsOpen(open);
    }
  }, [open]);

  return (
    <Portal>
      <ControlledMenu
        ref={menuRef}
        state={isOpen ? 'open' : 'closed'}
        align='center'
        direction='top'
        arrow
        className={classnames({ shake: isShaking })}
        anchorRef={anchorRef}
        skipOpen={skipOpen}>
        <div className='popover-container'>
          <div className='popover-title'>{title}</div>
          <div className='popover-buttons'>
            {buttons.map((button) => (
              <Button key={`${button.label}-${button.icon?.toString()}`} {...button} />
            ))}
          </div>
        </div>
      </ControlledMenu>
    </Portal>
  );
};

export default Popover;
