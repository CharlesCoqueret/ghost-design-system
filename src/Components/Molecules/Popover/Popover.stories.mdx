import { useState, useRef } from 'react';
import { Canvas, Meta, ArgsTable, Story } from '@storybook/addon-docs';

import Popover from './Popover';
import Button, { ButtonColorEnum } from '../Button/Button';

<Meta
  title='Molecule/Popover'
  component={Popover}
  parameters={{
    viewMode: 'docs',
    layout: 'centered',
  }}
  argTypes={{
    dataTestId: {
      table: {
        disable: true,
      },
    },
  }}
/>

# Popover

export const Template = () => {
  const [open, setOpen] = useState(false);
  const ref = useRef(null);
  return (
    <div ref={ref}>
      <Button
        label='Click me to see the popover'
        icon={['fal', 'trash-alt']}
        onClick={() => {
          setOpen(true);
        }}
        color={ButtonColorEnum.REVERSED}
      />
      <Popover
        anchorRef={ref}
        buttons={[
          {
            label: 'Cancel',
            color: ButtonColorEnum.SECONDARY,
            onClick: () => {
              setOpen(false);
            },
          },
          {
            label: 'Confirm',
            color: ButtonColorEnum.PRIMARY,
            onClick: () => {
              setOpen(false);
            },
          },
        ]}
        open={open}
        onClose={() => {
          setOpen(false);
        }}
        title='Delete?'
      />
    </div>
  );
};

<Canvas withSource='open'>
  <Story name='Popover'>{Template.bind({})}</Story>
</Canvas>

<ArgsTable of={Popover} sort='requiredFirst' exclude={['dataTestId']} />
