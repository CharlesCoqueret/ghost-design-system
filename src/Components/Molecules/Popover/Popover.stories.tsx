import React, { useRef, useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Popover, { IPopoverProps } from './Popover';
import Button, { ColorButtonEnum } from '../Button/Button';

export default {
  title: 'Molecule/Popover',
  component: Popover,
  parameters: { actions: { argTypesRegex: '^on.*' }, controls: { sort: 'requiredFirst' }, layout: 'centered' },
} as ComponentMeta<typeof Popover>;

const Template: ComponentStory<typeof Popover> = (args: IPopoverProps) => {
  const [open, setOpen] = useState(false);
  const ref = useRef<HTMLDivElement>(null);

  return (
    <div ref={ref}>
      <Button
        label='Click me to see the popover'
        icon={['fal', 'trash-alt']}
        onClick={() => {
          setOpen(true);
        }}
        color={ColorButtonEnum.REVERSED}
      />
      <Popover
        {...args}
        anchorRef={ref}
        buttons={[
          {
            label: 'Cancel',
            color: ColorButtonEnum.SECONDARY,
            onClick: () => {
              console.log('Cancelled');
              setOpen(false);
            },
          },
          {
            label: 'Confirm',
            color: ColorButtonEnum.PRIMARY,
            onClick: () => {
              console.log('Confirmed');
              setOpen(false);
            },
          },
        ]}
        open={open}
        onClose={() => {
          console.log('Cancelled because of click outside');
          setOpen(false);
        }}
        title='Delete?'
      />
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {};
