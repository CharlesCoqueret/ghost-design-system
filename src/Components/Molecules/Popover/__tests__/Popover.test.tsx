import React, { ReactElement, useRef } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Popover from '../Popover';
jest.unmock('@szhsin/react-menu');

describe('Popover Component', () => {
  afterAll(() => {
    jest.resetAllMocks();
  });

  // Creating wrapper component to be able to access useRef
  const MockComponent = ({ onCloseMock, open }: { onCloseMock: () => void; open: boolean }): ReactElement => {
    const ref = useRef(null);
    return (
      <>
        <div ref={ref} />
        <Popover anchorRef={ref} buttons={[]} onClose={onCloseMock} open={open} title='' />
      </>
    );
  };

  it('Popover renders', async () => {
    const onCloseMock = jest.fn();
    const { baseElement, rerender } = render(<MockComponent onCloseMock={onCloseMock} open={false} />);

    expect(baseElement).toMatchSnapshot();

    rerender(<MockComponent onCloseMock={onCloseMock} open={true} />);
    expect(baseElement).toMatchSnapshot();

    await userEvent.click(document.body);
    expect(onCloseMock).toBeCalledTimes(1);
  });
});
