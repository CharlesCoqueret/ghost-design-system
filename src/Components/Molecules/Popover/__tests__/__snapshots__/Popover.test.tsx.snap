// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Component Popover renders 1`] = `
<body>
  <div>
    <div />
    <div>
      <div
        class="szh-menu-container"
        style="position: absolute;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-closed szh-menu--dir-top"
          role="menu"
          style="margin: 0px; display: none; position: absolute; left: -9999px; top: -9999px;"
          tabindex="-1"
        >
          <li
            aria-hidden="true"
            role="menuitem"
            style="position: absolute; left: 0px; top: 0px; display: block; outline: none;"
            tabindex="-1"
          />
          <li
            aria-hidden="true"
            class="szh-menu__arrow szh-menu__arrow--dir-top"
            role="menuitem"
            style="display: block; position: absolute;"
          />
          <div
            class="container"
          >
            <div
              class="title"
            />
            <div
              class="buttons"
            />
          </div>
        </ul>
      </div>
    </div>
  </div>
  <div
    id="popover-portal-id"
  >
    <div />
  </div>
</body>
`;

exports[`Popover Component Popover renders 2`] = `
<body>
  <div>
    <div />
    <div>
      <div
        class="szh-menu-container"
        style="position: absolute;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-open szh-menu--dir-top"
          role="menu"
          style="margin: 0px; position: absolute; left: 0px; top: 0px;"
          tabindex="-1"
        >
          <li
            aria-hidden="true"
            role="menuitem"
            style="position: absolute; left: 0px; top: 0px; display: block; outline: none;"
            tabindex="-1"
          />
          <li
            aria-hidden="true"
            class="szh-menu__arrow szh-menu__arrow--dir-top"
            role="menuitem"
            style="display: block; position: absolute; left: 0px;"
          />
          <div
            class="container"
          >
            <div
              class="title"
            />
            <div
              class="buttons"
            />
          </div>
        </ul>
      </div>
    </div>
  </div>
  <div
    id="popover-portal-id"
  >
    <div />
  </div>
</body>
`;
